File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        map = {'(':')', '[':']', '{':'}'}
        stack = []
        for p in s:
            if p in map:
                stack.append(p)
            else:
                if len(stack) == 0 or map[stack.pop()] != p:
                    return False
        return True if len(stack) == 0 else False

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        HashMap<Character, Character> bracketsMap = new HashMap<>();
        bracketsMap.put(')', '(');
        bracketsMap.put(']', '[');
        bracketsMap.put('}', '{');
        Stack<Character> st = new Stack<>();
        st.push(s.charAt(0));
        for (int i = 1; i < s.length(); i++) {
            char b = s.charAt(i);
            if (bracketsMap.containsKey(b)) {
                if (st.isEmpty() || st.peek() != bracketsMap.get(b)) return false;
                st.pop();
            } else {
                st.push(b);
            }
        }
        return st.isEmpty();
    }
}